src/FffGcodeWriter.cpp

```diff
@@ -1425,155 +1425,237 @@ bool FffGcodeWriter::processSkinAndPerimeterGaps(const SliceDataStorage& storage
             }
         }
-
-        // determine inner_skin_outline
-        if (skin_part.insets.size() > 0)
-        {
-            inner_skin_outline = &skin_part.insets.back();
-            offset_from_inner_skin_outline = -mesh_config.insetX_config.getLineWidth() / 2;
-        }
     }
+}
 
-    if (inner_skin_outline == nullptr)
+void FffGcodeWriter::processSkinPartTopMostInfillGeneratePerimeterGaps(const SliceDataStorage& storage, LayerPlan& gcode_layer, const SliceMeshStorage& mesh, const int extruder_nr, const PathConfigStorage::MeshPathConfigs& mesh_config, const SkinPart& skin_part, int skin_angle, const bool generate_perimeter_gaps, Polygons& concentric_perimeter_gaps, bool& added_something) const
+{
+    coord_t skin_overlap = 0; // skinfill already expanded over the topmost skin; don't overlap with perimeters
+    EFillMethod pattern = mesh.getSettingAsFillMethod("topmost_skin_pattern");
+
+    const coord_t skin_line_width = mesh_config.topmost_skin_config.getLineWidth();
+    const int topmost_skin_extruder_nr = mesh.getSettingAsExtruderNr("topmost_skin_extruder_nr");
+
+    Polygons skin_polygons;
+    Polygons skin_lines;
+
+    const bool process_fill = extruder_nr == topmost_skin_extruder_nr;
+
+    // generate skin_polygons and skin_lines (and concentric_perimeter_gaps if needed)
+    if (process_fill || generate_perimeter_gaps)
     {
-        inner_skin_outline = &skin_part.outline;
-    }
+        // only generate infill if we're going to print it with this extruder
+        // or when we need to compute the perimeter gaps
+        // TODO: only compute this once when the infill is concentric and the perimeter gpas are printed with a different extruder
 
-    int extra_infill_shift = 0;
-    Polygons concentric_perimeter_gaps; // the perimeter gaps of the insets of concentric skin pattern of this skin part
-    Polygons* perimeter_gaps_output = (fill_concentric_perimeter_gaps)? &concentric_perimeter_gaps : nullptr;
-    Infill infill_comp(pattern, *inner_skin_outline, offset_from_inner_skin_outline, skin_line_width, skin_line_width, skin_overlap, skin_angle, z, extra_infill_shift, perimeter_gaps_output);
-    infill_comp.generate(skin_polygons, skin_lines);
+        // calculate polygons and lines
+        constexpr int extra_infill_shift = 0;
+        constexpr coord_t offset_from_inner_skin_infill = 0;
+        Polygons* perimeter_gaps_output = (generate_perimeter_gaps)? &concentric_perimeter_gaps : nullptr;
+        Infill infill_comp(pattern, skin_part.top_most_skinfill, offset_from_inner_skin_infill, skin_line_width, skin_line_width, skin_overlap, skin_angle, gcode_layer.z, extra_infill_shift, perimeter_gaps_output);
+        infill_comp.generate(skin_polygons, skin_lines);
+    }
 
     // add skin itself!
-    if ((skin_polygons.size() > 0 || skin_lines.size() > 0) && extruder_nr == top_bottom_extruder_nr)
+    if (process_fill && (skin_polygons.size() > 0 || skin_lines.size() > 0))
```

Comment:
No. The previous clause has different conditions, i.e. `|| generate_perimeter_gaps`